### PR TITLE
107 mixed space adj project

### LINKED ISSUE #107: Issue when using SWE mixed function space - checking `project` function source
@jwallwork23  - I hit an error when running goal-based adaptation with two time-partitions with a SWE mixed-function space. The error originated from firedrake `project` and I noticed you have implemented an override of `project` in goalie. 

Is the `project` implemented here setup to handle mixed function spaces?

https://github.com/pyroteus/goalie/blob/472e496d284ad0e8db4b6f1076bf29504adc8b7e/goalie/adjoint.py#L383-L387

### PR DESCRIPTION
**Pull request by acse-ej321**
fixes #107 
Calls the override `project` function instead of the default Firedrake function to deal with mixed function spaces and adjoints.

Additionally, modifies the creation of the enriched mesh to:
- pass inherited parameters correctly in the constructor
- force population of `self._fs` variable required for mesh-to-mesh conservative transfer

### CODE REVIEW

File: goalie/go_mesh_seq.py [hunk 1]
```diff
@@ -61,8 +62,12 @@ def get_enriched_mesh_seq(
             get_qoi=self._get_qoi,
             get_bcs=self._get_bcs,
             qoi_type=self.qoi_type,
+            parameters=self.params,
```
Comment (jwallwork23) [inline goalie/go_mesh_seq.py:65]: Thanks. Partially addresses #97.

File: goalie/go_mesh_seq.py [hunk 2]
```diff
@@ -61,8 +62,12 @@ def get_enriched_mesh_seq(
...
         )
 
+        # update self._fs variable for enriched mesh
+        mesh_seq_e.function_spaces
```
Comment (jwallwork23) [inline goalie/go_mesh_seq.py:69]: Please could you introduce a method on the `MeshSeq` to do this whose name says what it is doing, e.g., `update_function_spaces`? On it's own, this line of code looks like it's doing nothing. (My bad for writing it this way in the first place, I was probably overusing properties.)

**Review by jwallwork23 — CHANGES_REQUESTED** [review]
Thanks for this. Seems the right thing to do.

One request and then we'll need to merge #104 to get fixes that are currently holding up the test suite.

File: goalie/go_mesh_seq.py (hunk 2 above)
Comment (acse-ej321) [reply goalie/go_mesh_seq.py]: @jwallwork23 
I added a function called `update_function_spaces` which calls `function_spaces` if the function space attribute dictionary is `None`. Hopefully, that is a bit more readable.

File: goalie/go_mesh_seq.py [hunk 3]
```diff
@@ -64,9 +64,7 @@ def get_enriched_mesh_seq(
             qoi_type=self.qoi_type,
             parameters=self.params,
         )
-
-        # Update function spaces for enriched mesh
-        mesh_seq_e.update_function_spaces
```
Comment (jwallwork23) [inline goalie/go_mesh_seq.py]: Note that I prefer not to add comments that are already clear from the code / method name.

Comment (jwallwork23) [reply goalie/go_mesh_seq.py]: I'm writing some [Development Practices](https://github.com/pyroteus/movement/wiki/Development-Practices) over in the Movement wiki and I'll put this in there.

Comment (acse-ej321) [reply goalie/go_mesh_seq.py]: @jwallwork23 - Noted. I have a tendency to over-comment and will work on it especially when the attempt is to make the method name self explanatory.

File: goalie/mesh_seq.py [hunk 4]
```diff
@@ -234,8 +235,10 @@ def _function_spaces_consistent(self) -> bool:
             )
         return consistent
 
-    @property
-    def function_spaces(self) -> list:
+    def update_function_spaces(self) -> list:
```
Comment (acse-ej321) [inline goalie/mesh_seq.py:238]: @jwallwork23 - I think your edits here make more sense than my interpretation of your suggestions and it is a better descriptive name of the method you had already implemented.

**Review by jwallwork23 — APPROVED** [review]
Thanks again for this @acse-ej321!

**Comment by jwallwork23** [comment]
@acse-ej321 Are you happy to merge this? It's good to go. (Usually I leave merging to the developer.)

**Pull request merged**